clients/da-vinci-client/src/main/java/com/linkedin/davinci/replication/merge/MergeConflictResolver.java

```diff
@@ -312,15 +312,14 @@ private MergeConflictResult mergePutWithFieldLevelTimestamp(
       long newValueSourceOffset,
       int newValueSourceBrokerID,
       int newValueSchemaID) {
-    if (!(oldTimestampObject instanceof GenericRecord)) {
-      throw new IllegalStateException(
-          "Per-field RMD timestamp must be a GenericRecord. Got: " + oldTimestampObject + " and store name is: "
-              + storeName);
-    }
-    final GenericRecord oldValueFieldTimestampsRecord = (GenericRecord) oldTimestampObject;
-    if (ignoreNewPut(oldValueSchemaID, oldValueFieldTimestampsRecord, newValueSchemaID, putOperationTimestamp)) {
-      return MergeConflictResult.getIgnoredResult();
+
+    if (oldTimestampObject instanceof GenericRecord) {
+      final GenericRecord oldValueFieldTimestampsRecord = (GenericRecord) oldTimestampObject;
+      if (ignoreNewPut(oldValueSchemaID, oldValueFieldTimestampsRecord, newValueSchemaID, putOperationTimestamp)) {
+        return MergeConflictResult.getIgnoredResult();
+      }
```

Comment:
Just by the virtue of not throwing an exception for non generic per-field RMD, sufficient for converting the value level TS to field level TS?

I don't have a clear picture on what takes precedence and how one overrides another by removing this check. Feel free to point me to code blocks that explain the behavior.

Reply:
Yes, there is a code below to actually perform auto conversion:
createOldValueAndRmd in which it calls:
    // RMD record should contain a per-field timestamp and it should use the RMD schema generated from
    // mergeResultValueSchema.
    oldRmdRecord = convertToPerFieldTimestampRmd(oldRmdRecord, oldValueRecord);